promql/parse.go

```diff
@@ -0,0 +1,868 @@
+		Name:   name,
+		Labels: lset,
+		Expr:   expr,
+	}
+}
+
+// expr parses any expression.
+func (p *parser) expr() Expr {
+	const ctx = "binary expression"
+
+	// Parse the starting expression.
+	expr := p.unaryExpr()
+
+	// Loop through the operations and construct a binary operation tree based
+	// on the operators' precedence.
+	for {
+		// If the next token is not an operator the expression is done.
+		op := p.peek().typ
+		if !op.isOperator() {
+			return expr
+		}
+		p.next() // Consume operator.
+
+		// Parse optional operator matching options. Its validity
+		// is checked in the type-checking stage.
+		vecMatching := &VectorMatching{
+			Card: CardOneToOne,
+		}
+		if op == itemLAND || op == itemLOR {
+			vecMatching.Card = CardManyToMany
+		}
+
+		// Parse ON clause.
+		if p.peek().typ == itemOn {
+			p.next()
+			vecMatching.On = p.labels()
+
+			// Parse grouping.
+			if t := p.peek().typ; t == itemGroupLeft {
+				p.next()
+				vecMatching.Card = CardManyToOne
+				vecMatching.Include = p.labels()
+			} else if t == itemGroupRight {
+				p.next()
+				vecMatching.Card = CardOneToMany
+				vecMatching.Include = p.labels()
+			}
+		}
+
+		for _, ln := range vecMatching.On {
+			for _, ln2 := range vecMatching.Include {
+				if ln == ln2 {
+					p.errorf("label %q must not occur in ON and INCLUDE clause at once", ln)
+				}
+			}
+		}
+
+		// Parse the next operand.
+		rhs := p.unaryExpr()
+
+		// Assign the new root based on the precendence of the LHS and RHS operators.
+		if lhs, ok := expr.(*BinaryExpr); ok && lhs.Op.precedence() < op.precedence() {
+			expr = &BinaryExpr{
+				Op:  lhs.Op,
+				LHS: lhs.LHS,
+				RHS: &BinaryExpr{
+					Op:             op,
+					LHS:            lhs.RHS,
+					RHS:            rhs,
+					VectorMatching: vecMatching,
+				},
+				VectorMatching: lhs.VectorMatching,
+			}
+		} else {
+			expr = &BinaryExpr{
+				Op:             op,
+				LHS:            expr,
+				RHS:            rhs,
+				VectorMatching: vecMatching,
+			}
+		}
+	}
+	return nil
+}
+
+// unaryExpr parses a unary expression.
+//
+//		<vector_selector> | <matrix_selector> | (+|-) <number_literal> | '(' <expr> ')'
+//
+func (p *parser) unaryExpr() Expr {
+	switch t := p.peek(); t.typ {
+	case itemADD, itemSUB:
+		p.next()
+		e := p.unaryExpr()
+		// Simplify unary expressions for number literals.
+		if nl, ok := e.(*NumberLiteral); ok {
+			if t.typ == itemSUB {
+				nl.Val *= -1
+			}
+			return nl
+		}
+		return &UnaryExpr{Op: t.typ, Expr: e}
+
+	case itemLeftParen:
+		p.next()
+		e := p.expr()
+		p.expect(itemRightParen, "paren expression")
+
+		return &ParenExpr{Expr: e}
+	}
+	e := p.primaryExpr()
+
+	// Expression might be matrix selected.
+	if p.peek().typ == itemLeftBracket {
+		vs, ok := e.(*VectorSelector)
+		if !ok {
+			p.errorf("matrix selected expression must be a metric selection but is %T", e)
+		}
+		e = p.matrixSelector(vs)
+	}
+	return e
+}
+
+// matrixSelector parses a matrix selector based on a given vector selector.
```

Comment:
Same thing here. I wouldn't call it a "matrix selector" (the matrix is the result of both the metric selector and the range selector), but either a "range selector" or a "range specification", whichever you prefer.